runner/src/main.rs

```diff
@@ -23,16 +23,152 @@
+};
 use structopt::StructOpt;
 
 mod internal;
 use internal::*;
 
 fn main() -> Result<(), Box<dyn std::error::Error>> {
     let opt = Opt::from_args();
+
+    let steps = match opt.cmd {
+        Command::BuildExamples(ref opt) => build_examples(&opt),
+        Command::BuildServer(ref opt) => build_server(&opt),
+        Command::RunTests => run_tests(),
+    };
     // TODO(#396): Add support for running individual commands via command line flags.
-    let root = Step::Multiple {
+    let statuses = run_step(&Context::root(&opt), &steps);
+
+    // If the overall status value is an error, terminate with a nonzero exit code.
+    if statuses.contains(&StatusResultValue::Error) {
+        std::process::exit(1);
+    }
+
+    Ok(())
+}
+
+fn build_examples(opt: &BuildExamples) -> Step {
+    match opt.variant.as_str() {
+        "rust" => Step::Multiple {
+            name: "root".to_string(),
+            steps: examples()
+                .filter(is_cargo_toml_file)
+                .filter(|entry| {
+                    (*entry != PathBuf::from("./examples/aggregator/backend/Cargo.toml"))
+                        && (*entry != PathBuf::from("./examples/authentication/server/Cargo.toml"))
+                        && (*entry != PathBuf::from("./examples/authentication/client/Cargo.toml"))
+                })
+                .map(to_string)
+                .map(|entry| build_wasm_module(&entry))
+                .collect(),
+        },
+        "cpp" => unimplemented!("C++ examples not implemented yet"),
+        v => panic!("unknown variant: {}", v),
+    }
+}
+
+fn build_wasm_module(manifest_path: &str) -> Step {
+    Step::Single {
+        name: manifest_path.to_string(),
+        runnable: cmd(
+            "cargo",
+            &[
+                "build",
+                "--release",
+                "--target=wasm32-unknown-unknown",
+                &format!("--manifest-path={}", manifest_path),
+            ],
+        ),
+    }
+}
+
+fn is_mac_os() -> bool {
+    cfg!(macos)
+}
+
+fn bazel_build_flags() -> Vec<String> {
+    let mut flags = vec![
+        "--remote_cache=https://storage.googleapis.com/oak-bazel-cache".to_string(),
+        "--block_for_lock=false".to_string(),
+        "--show_timestamps".to_string(),
+    ];
+
+    const OAK_REMOTE_CACHE_KEY: &str = "./.oak_remote_cache_key.json";
+    if Path::new(OAK_REMOTE_CACHE_KEY).exists() {
+        flags.push(format!("--google_credentials={}", OAK_REMOTE_CACHE_KEY));
+    } else {
+        flags.push("--remote_upload_local_results=false".to_string());
+    };
+
+    flags
+}
+
+fn as_ref(v: &Vec<String>) -> Vec<&str> {
+    v.iter().map(|x| x.as_ref()).collect()
+}
+
+fn bazel_build(extra_build_flags: &[&str], targets: &[&str]) -> Box<dyn Runnable> {
+    let mut flags = Vec::new();
+    flags.push("build");
+
+    let b = bazel_build_flags();
+    flags.extend(as_ref(&b));
+
+    flags.extend(extra_build_flags);
+    flags.push("--");
+    flags.extend(targets);
```

Comment:
Do we need to also support excluded targets?

Reply:
No need for now (also Bazel will be even less relevant going forward)